drivers/saul/gpio_saul.c

```diff
@@ -23,12 +23,16 @@
 #include "saul.h"
 #include "phydat.h"
 #include "periph/gpio.h"
+#include "saul/periph.h"
 
 
 static int read(const void *dev, phydat_t *res)
 {
-    gpio_t pin = *((const gpio_t *)dev);
-    res->val[0] = (gpio_read(pin)) ? 1 : 0;
+    const saul_gpio_params_t *p = (const saul_gpio_params_t *)dev;
+    int inverted = (p->flags | SAUL_GPIO_INVERTED);
+
+    res->val[0] = (gpio_read(p->pin)) ? !inverted : inverted;
```

Comment:
I think you could use XOR here and below, i.e.,

```
res->val[0] = gpio_read(p->pin) ^ inverted;
```

Reply:
I am doing an XOR, but not a bitwise xor.

I just checked and no it cannot be replaced, `gpio_read` is only expected to return a positive value on HIGH. https://github.com/RIOT-OS/RIOT/blob/2017.07/drivers/include/periph/gpio.h#L189

When looking at actual implementation, the first I found in alpha order is https://github.com/RIOT-OS/RIOT/blob/2017.07/cpu/atmega_common/periph/gpio.c#L209 which in fact does not return 1.

The previous code was also doing this `(gpio_read(pin) ? 1 : 0)`
The thing I have here, is that I expect `SAUL_GPIO_INVERTED` to be equal to `1`. Maybe it would require a static assert for this.